app/services/budgetkey-item.ts

```diff
@@ -46,22 +46,42 @@ export class BudgetKeyItemService {
     Promise.reject(new Error('No layout for ' + path));
   }
 
-  getItemData(query: string, headersOrder: string[]): Promise<object> {
-    let url = 'http://next.obudget.org/api/query?query=' +
-      encodeURIComponent(query);
+  private _budgetNumberFormatter(value: any) {
+    return parseFloat(value).toLocaleString('he-IL', { style: 'currency', currency: 'ILS' });
+  }
+
+  private _budgetLinkFormatter(value: string, hLink: string) {
+    return '<a>' + value + '</a>';
+  }
+
+  getItemData(query: string, headersOrder: string[], formatters: string[]): Promise<object> {
+    const url = 'http://next.obudget.org/api/query?query=' + encodeURIComponent(query),
+    formatterKeys = _.keys(formatters),
+    formatterValues = _.values(formatters);
+
     return new Promise<any>((resolve, reject) => {
       this.http.get(url)
         .map((res: any) => res.json())
         .subscribe(
           (res: any) => {
-            let items: object[] = [];
-            let rows = res.rows;
-            let headers = rows.length > 0 ? _.union(headersOrder, _.keys(_.first(rows))) : [];
+            let items: object[] = [],
+            rows = res.rows,
+            headers = rows.length > 0 ? _.union(headersOrder, _.keys(_.first(rows))) : [];
+
             _.each(rows, (row) => {
               let newItem: any[] = [];
+
               _.each(headers, (header) => {
-                newItem.push(row[header]);
+                let item = row[header];
+
+                if (typeof item === 'string' || typeof item === 'number') {
+                  _.each(formatterKeys, (fKey, i) => {
```

Comment:
why not access formatters[header] here ??

Reply:
Sorry, probably that's not my business, but why not `_.each(formatters, (formatterValue, formatterKey) => { ... ` ?

Reply:
@kravets-levko, your comments are always welcome! 😃 
you're right of course - except that we don't need to iterate on an object's keys to find a specific key's value...

Reply:
Hm, well, I tried to read some lines above and below. As I understand, formatter is identified by column name. Then it's better to rewrite entire function in a way like this:
```javascript
const items = _.map(rows, (row) => {
  return _.map(headers, (header) => {
    const value = row[header];
    if (_.isString(value) || _isNumber(value)) {
      return formatters[header] ? formatters[header](value) : value;
    }
    return value;
  };
});
```

Reply:
Also we can make some optimization here: prepare all formatters before mapping rows - and remove all that `if`s inside (because they will run in two nested loops...):
```javascript
const preparedFormatters = {};
_.each(header, (header) => {
  preparedFormatters[header] = _.isFunction(this[formatters[header]]) ?
    this[formatters[header]].bind(this) : _.identity;
});

const items = _.map(rows, (row) => {
  return _.map(headers, (header) => {
    // No checks here because `preparedFormatters` will always contain all headers,
    // and all values will be functions
    return preparedFormatters[header](row[header]);
  };
});
```

Reply:
take a look at the last version and inform me if it good enough